python-api/livy/job_handle.py

```diff
@@ -0,0 +1,257 @@
+        """
+        with self._job_handle_condition:
+            return self._state == QUEUED
+
+    def _invoke_queued_callbacks(self):
+        for callback in self._queued_callbacks:
+            try:
+                callback(self)
+            except Exception:
+                traceback.print_exc()
+            self._queued_callbacks.remove(callback)
+
+    def _invoke_running_callbacks(self):
+        for callback in self._running_callbacks:
+            try:
+                callback(self)
+            except Exception:
+                traceback.print_exc()
+            self._running_callbacks.remove(callback)
+
+    def add_queued_callback(self, fn):
+        """
+        Attaches a callable that will be called when the job state is
+        queued.
+
+        Parameters
+        ----------
+        fn : Function
+            A callable that will be called with this future as its only
+            argument when the job state is queued. The
+            callable will always be called by a thread in the same
+            process in which it was added. These callables are called in
+            the order that they were added.
+        """
+        with self._job_handle_condition:
+            if self._state == PENDING:
+                self._queued_callbacks.append(fn)
+            elif self._state == QUEUED:
+                fn(self)
+
+    def add_running_callback(self, fn):
+        """
+        Attaches a callable that will be called when the job state is
+        running.
+
+        Parameters
+        ----------
+        fn : Function
+            A callable that will be called with this future as its only
+            argument when the job state is running. The
+            callable will always be called by a thread in the same
+            process in which it was added. These callables are called in
+            the order that they were added.
+        """
+        with self._job_handle_condition:
+            if self._state in [PENDING, QUEUED]:
+                self._running_callbacks.append(fn)
+            elif self._state == RUNNING:
+                fn(self)
+
+    def _update_state(self, state):
+        with self._job_handle_condition:
+            if state == 'STARTED':
+                self._state = RUNNING
+                self._invoke_running_callbacks()
+            elif state == 'QUEUED':
+                self._state = QUEUED
+                self._invoke_queued_callbacks()
+            elif state == 'CANCELLED':
+                self._state = CANCELLED
+                self._invoke_callbacks()
+            else:
+                raise RuntimeError('Future in unexpected state::', self._state)
+            self._job_handle_condition.notifyAll()
+
+    def cancel(self):
+        """
+        Cancel the future if possible.
+
+        Returns
+        -------
+        True if the future was cancelled, False otherwise. A future
+        cannot be cancelled if it is running or has already completed.
```

Comment:
It might be possible to cancel a running job. You don't know that from the client side. So you shouldn't make decisions about whether a job can or cannot be cancelled here (except if it has already been cancelled, or is finished). Just send a request to the server.

Java's `Future` (which Java's `JobHandle` extends) takes a parameter to this method, also, to say whether the running job should be interrupted.